### PR TITLE
Imagestream tag exclude from pruning - using object selection

### PR DESCRIPTION
**Pull request by deads2k**
Alternative to https://github.com/openshift/origin/pull/16580.

To help clarify comments in https://github.com/openshift/origin/pull/16580, https://bugzilla.redhat.com/show_bug.cgi?id=1410018, and https://github.com/openshift/api/pull/9 I have put together a complete pull which adds the ability to protect namespaces and/or individual imagestreamtags based on our standard label and field selection techniques.  This pull allows flows like this:

```
# old stock behavior
oc adm prune images <normal args>

# prune images, but only in unprotected namespaces (label of my choosing)
oc adm prune images <normal args> -l 'cluster-admin-protection-status notin (special)'

# prune images, but only gold level, then prune silver with different args
oc adm prune images <normal args> -l 'plan=gold'
oc adm prune images <more restrictive args> -l 'plan=silver'

# protect particular images, you don't get to choose this label.  myspecial:latest isn't pruned
oc label ns/openshift "image.openshift.io/allow-prune-immunity=true"
oc annotate -n openshift istag/myspecial:latest "image.openshift.io/prune-immunity=true"
oc adm prune images <normal args> [optional selector]
```

I'd still prefer to label the istag instead of using the equivalent of a field selector, but this gets the job done without any API changes.

@bparees @smarterclayton @mfojtik @openshift/cli-review

### CODE REVIEW

**Comment by bparees** [comment]
@deads2k are you going to write the docs for this approach (@legionus already wrote docs for the earlier implementation)?

Are you going to work w/ QE to update their tests and shepherd it through the rest of the process since the design has changed?

I know you think this is trivial, but you're creating significant churn on our side that you seem unwilling to acknowledge.

Also from what I see this approach does not currently offer a way to *ignore* immunity(let alone selectively ignore it), which is another advantage of the regex approach.

**Comment by deads2k** [comment]
> Also from what I see this approach does not currently offer a way to ignore immunity(let alone selectively ignore it), which is another advantage of the regex approach.

Ignoring immunity is essentially the same as removing a label or an annotation.  With either implementation it requires a modification to perform selectively.  To perform non-selectively, you can simply remove labels globally.  As with the rest of our API objects, the ACL boundary is the namespace.



> @deads2k are you going to write the docs for this approach (@legionus already wrote docs for the earlier implementation)?
> 
> Are you going to work w/ QE to update their tests and shepherd it through the rest of the process since the design has changed?

If need be, yes.  I would prefer to do that and gain the benefits described in the bugzilla along with consistency of experience when using the CLI.  

My preference would be for broad CLI changes introducing new patterns or deviating from existing patterns to be subject to review in the broad sense and individual commands to maintained separately.  This would mirror the approach upstream, where additions using existing patterns are subject to cursory review by sig-cli-leads and involved changes are more heavily scrutinized.  I don't think that juan or fabiano are pruning experts, but the experience they are trying to cultivate and maintain for consistency within `oc` is something they are aware of.

If I need to own the card end-to-end, I can do so.  However, I think we'd be better served in the end to grow the knowledge about common patterns and user experience across the teams which own particular commands.  Consistency in these areas makes using new commands easier and knowledge of the general patterns gives more confidence when adding CLI features.

**Comment by bparees** [comment]
> Ignoring immunity is essentially the same as removing a label or an annotation. With either implementation it requires a modification to perform selectively. To perform non-selectively, you can simply remove labels globally. As with the rest of our API objects, the ACL boundary is the namespace.

i still don't agree that finding and updating N server side objects is easier or better than tweaking a few regexes locally (or not passing the regex at all to disable immunity).

**Comment by miminar** [comment]
> i still don't agree that finding and updating N server side objects is easier or better than tweaking a few regexes locally (or not passing the regex at all to disable immunity).

@bparees without the immunity being set on namespaces/imagestreams, each time the user attempts to prune manually, he needs to remember all namespaces/istags he wants to preserve and write it down into a regexp on command-line, which gets tedious. I know that being able to write this regexps into a file and passing it to the command helps, but the user still needs to remember he has some namespaces/istags he wants to protect.

IMHO the namespace immunity won't change easily over time - company will want particular namespaces/istags protect against pruning in any case. Setting the label once and not needing to care about it anymore is a much more convenient solution to the problem.

**Comment by miminar** [comment]
Building upon this approach also allows for per namespace/is pruning options. E.g. keep this X tag revisions in proj/myapp and X+2 revisions in proj/myprecious.

File: pkg/oc/admin/prune/images.go [hunk 1]
```diff
@@ -104,19 +106,24 @@ var (
 	defaultKeepYoungerThan         = 60 * time.Minute
 	defaultKeepTagRevisions        = 3
 	defaultPruneImageOverSizeLimit = false
+
+	// TODO choose final name and promote to API
+	protectionLabel = "image.openshift.io/allow-prune-immunity"
```
Comment (miminar) [inline pkg/oc/admin/prune/images.go]: Can it be more general? (e.g. `openshift.io/allow-prune-immunity`). We have pruners for resources other than image.

Comment (deads2k) [reply pkg/oc/admin/prune/images.go]: > Can it be more general? (e.g. openshift.io/allow-prune-immunity). We have pruners for resources other than image.

I would rather start specific and move out to generality.  Pruning isn't a well explored space yet.  See how deployments do it as a counter example.

File: pkg/image/prune/prune.go [hunk 2]
```diff
@@ -351,8 +365,18 @@ func (p *pruner) addImageStreamsToGraph(streams *imageapi.ImageStreamList, limit
 					continue
 				}
 
+				isNamespaceEligible := p.namespacesToPrune.Has(stream.Namespace)
+				doesNamespaceAllowImmunity := p.namespacesToAllowImmunity.Has(stream.Namespace)
+				isISTagImmune := stream.Spec.Tags[tag].Annotations[protectionAnnotation] == "true"
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/image/prune/prune.go]: > I'd welcome per-imagestream control as well. Where istag's annotation overrides is' annotation.

It's easy enough to add, but the initial use-case didn't request it.  @miminar could it be a followup?

@soltysh @fabianofranz @juanvallejo add it or no?

File: pkg/oc/admin/prune/images.go [hunk 3]
```diff
@@ -162,6 +170,7 @@ func NewCmdPruneImages(f *clientcmd.Factory, parentName, name string, out io.Wri
...
 	cmd.Flags().StringVar(&opts.RegistryUrlOverride, "registry-url", opts.RegistryUrlOverride, "The address to use when contacting the registry, instead of using the default value. This is useful if you can't resolve or reach the registry (e.g.; the default is a cluster-internal URL) but you do have an alternative route that works. Particular transport protocol can be enforced using '<scheme>://' prefix.")
 	cmd.Flags().BoolVar(&opts.ForceInsecure, "force-insecure", opts.ForceInsecure, "If true, allow an insecure connection to the docker registry that is hosted via HTTP or has an invalid HTTPS certificate. Whenever possible, use --certificate-authority instead of this dangerous option.")
+	cmd.Flags().StringVarP(&opts.NamespacesToPruneSelector, "selector", "l", "", "Selector (label query) to filter on, not including uninitialized ones, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2).")
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/oc/admin/prune/images.go]: > I would turn the e.g. part into a full example.

Not sure I understand.  Those are valid examples.

**Comment by deads2k** [comment]
> Building upon this approach also allows for per namespace/is pruning options. E.g. keep this X tag revisions in proj/myapp and X+2 revisions in proj/myprecious.

Yep.

**Comment by deads2k** [comment]
@soltysh @miminar I have rebased.

I like the suggestion of respecting the flag set on an imagestream, but I'd like to move the bulk of the work and circle back to it.  Perhaps @juanvallejo could pick it up and work through the test matrix?

**Comment by deads2k** [comment]
https://github.com/openshift/origin/issues/17879
https://github.com/openshift/origin/issues/17880

File: pkg/image/prune/prune.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/image/prune/prune.go]: 👍 for adding it (can be followup). IIRC discussions in the topic it was about both IS and ISTag to be excluded from pruning.

File: pkg/image/prune/prune.go [hunk 4]
```diff
@@ -55,6 +55,9 @@ const (
 	// ReferencedImageLayerEdgeKind defines an edge from an ImageStreamNode or an
 	// ImageNode to an ImageComponentNode.
 	ReferencedImageLayerEdgeKind = "ReferencedImageLayer"
+
+	// TODO choose final name and promote to API
+	protectionAnnotation = "image.openshift.io/prune-immunity"
```
Comment (soltysh) [inline pkg/image/prune/prune.go]: @openshift/api-review ptal (yes David I know it's you, but also others 😉)

**Comment by soltysh** [comment]
The implementation itself lgtm, I'm ok with @juanvallejo working on the followup. Just need to make it in a working state.

File: pkg/image/prune/prune.go (hunk 4 above)
Comment (deads2k) [reply pkg/image/prune/prune.go]: > @openshift/api-review ptal (yes David I know it's you, but also others )

Fair. :)

File: pkg/image/prune/prune.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/image/prune/prune.go]: :+1: for a follow-up

**Comment by mfojtik** [comment]
@deads2k digged into the failed test, I think all 4 image streams that failed to update have one thing in common. They were all `docker push-ed` and the image stream tag was created by registry...

I wonder if we default the `referencePolicy.type` in API or if registry can somehow bypass this validation check. 

I guess this is a separate issue, so we can opt-out the failing image streams (`grep -v` or something) for now to move this PR forward. Make an issue for the @openshift/sig-developer-experience or @smarterclayton (who added referencePolicy) to investigate how it is possible that those tag have that field set to an empty string.

**Comment by mfojtik** [comment]
@bparees @openshift/sig-developer-experience were there any progress on determining the root cause of why certain image streams cannot be annotated (iow. we store the corrupted object in etcd) ?

**Comment by dmage** [comment]
@mfojtik #18519

**Comment by bparees** [comment]
> @bparees @openshift/sig-developer-experience were there any progress on determining the root cause of why certain image streams cannot be annotated (iow. we store the corrupted object in etcd) ?

@mfojtik it has always been this way, see my comment here where this was found and "worked around" (by setting the reference policy explicitly):
https://github.com/openshift/origin/pull/12731#pullrequestreview-97356755

it's effectively by design, see the entire discussion (including comment from clayton) here:
https://github.com/openshift/origin/issues/18519  (I tagged you in this issue, btw, are you not getting my notifications?  You haven't responded to a number of them)

and my PR to at least make the error message clearer:
https://github.com/openshift/origin/pull/18532

**Pull request closed without merging**